File: backend/app/routers/students.py
from fastapi import APIRouter, HTTPException, Depends
from app.database import get_supabase
from app.models.schemas import StudentProfileCreate, CreditRequestCreate
from app.dependencies import verify_token

router = APIRouter(prefix="/api/students", tags=["students"])


@router.get("/profile")
async def get_profile(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    result = supabase.table("student_profiles").select("*").eq("user_id", auth["user_id"]).execute()
    if not result.data:
        return {"success": True, "data": None}
    return {"success": True, "data": result.data[0]}


@router.put("/profile")
async def update_profile(profile: StudentProfileCreate, auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    existing = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
    if existing.data:
        result = supabase.table("student_profiles").update(profile.model_dump(exclude_none=True)).eq("user_id", auth["user_id"]).execute()
    else:
        result = supabase.table("student_profiles").insert({
            "user_id": auth["user_id"],
            **profile.model_dump(exclude_none=True)
        }).execute()
    return {"success": True, "data": result.data[0]}


@router.post("/credit-request")
async def submit_credit_request(req: CreditRequestCreate, auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    students = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
    if not students.data:
        raise HTTPException(status_code=400, detail="Complete your profile first")
    student = students.data[0]
    active = supabase.table("credit_applications").select("id").eq("student_id", student["id"]).in_("status", ["pending", "under_review", "scoring_done", "matched", "contract_proposed", "active"]).execute()
    if active.data:
        raise HTTPException(status_code=400, detail="You already have an active application")
    result = supabase.table("credit_applications").insert({
        "student_id": student["id"],
        "requested_monthly_amount": req.requested_monthly_amount,
        "requested_duration_months": req.requested_duration_months,
        "justification": req.justification,
        "status": "pending",
    }).execute()
    return {"success": True, "data": result.data[0]}


@router.get("/credit-request")
async def get_credit_request(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    students = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
    if not students.data:
        return {"success": True, "data": None}
    result = supabase.table("credit_applications").select("*").eq("student_id", students.data[0]["id"]).order("created_at", desc=True).limit(1).execute()
    return {"success": True, "data": result.data[0] if result.data else None}


@router.get("/documents")
async def list_documents(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    result = supabase.table("documents").select("*").eq("uploaded_by", auth["user_id"]).execute()
    return {"success": True, "data": result.data}


@router.get("/contract")
async def get_contract(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    students = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
    if not students.data:
        return {"success": True, "data": None}
    result = supabase.table("contracts").select("*").eq("student_id", students.data[0]["id"]).order("created_at", desc=True).limit(1).execute()
    return {"success": True, "data": result.data[0] if result.data else None}


@router.post("/contract/{contract_id}/accept")
async def accept_contract(contract_id: str, auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    result = supabase.table("contracts").update({"status": "accepted", "accepted_at": "now()"}).eq("id", contract_id).execute()
    return {"success": True, "data": result.data[0]}


@router.post("/contract/{contract_id}/refuse")
async def refuse_contract(contract_id: str, auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    result = supabase.table("contracts").update({"status": "terminated"}).eq("id", contract_id).execute()
    supabase.table("credit_applications").update({"status": "contract_refused"}).eq("id", result.data[0]["application_id"]).execute()
    return {"success": True, "data": result.data[0]}


@router.get("/payments")
async def list_payments(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    students = supabase.table("student_profiles").select("id").eq("user_id", auth["user_id"]).execute()
    if not students.data:
        return {"success": True, "data": []}
    contracts = supabase.table("contracts").select("id").eq("student_id", students.data[0]["id"]).execute()
    ids = [c["id"] for c in contracts.data] if contracts.data else []
    if not ids:
        return {"success": True, "data": []}
    result = supabase.table("payments").select("*").in_("contract_id", ids).order("due_date", desc=True).execute()
    return {"success": True, "data": result.data}


@router.get("/notifications")
async def list_notifications(auth: dict = Depends(verify_token)):
    supabase = get_supabase()
    result = supabase.table("notifications").select("*").eq("user_id", auth["user_id"]).order("created_at", desc=True).limit(20).execute()
    return {"success": True, "data": result.data}

File: backend/src/routes/students.ts
import { Router, Request, Response } from "express";
import { verifyToken } from "../middleware/auth";
import { supabase } from "../database";
import type { StudentProfileInput, CreditRequestInput } from "../types";

const router = Router();

async function tryUpsertProfile(data: Record<string, any>, userId: string) {
  const { data: existing } = await supabase
    .from("student_profiles")
    .select("id")
    .eq("user_id", userId);
  if (existing && existing.length > 0) {
    const { data: result } = await supabase
      .from("student_profiles")
      .update(data)
      .eq("user_id", userId)
      .select();
    return result;
  }
  const { data: result } = await supabase
    .from("student_profiles")
    .insert({ user_id: userId, ...data })
    .select();
  return result;
}

router.get("/profile", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data, error } = await supabase
      .from("student_profiles")
      .select("*")
      .eq("user_id", req.user!.user_id);
    if (error) return res.status(400).json({ detail: error.message });
    return res.json({ success: true, data: data && data.length > 0 ? data[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.put("/profile", verifyToken, async (req: Request, res: Response) => {
  try {
    const body: StudentProfileInput = req.body;
    const data = Object.fromEntries(
      Object.entries(body).filter(([_, v]) => v !== null && v !== undefined)
    );
    let result;
    try {
      result = await tryUpsertProfile(data, req.user!.user_id);
    } catch (err: any) {
      if (
        err.message &&
        (err.message.includes("education_level") || err.message.includes("extra_data"))
      ) {
        const safeData = { ...data };
        delete safeData.education_level;
        delete safeData.extra_data;
        result = await tryUpsertProfile(safeData, req.user!.user_id);
      } else {
        throw err;
      }
    }
    return res.json({ success: true, data: result ? result[0] : null });
  } catch (err: any) {
    return res.status(400).json({ detail: err.message });
  }
});

router.post("/credit-request", verifyToken, async (req: Request, res: Response) => {
  try {
    const body: CreditRequestInput = req.body;
    const { data: students, error: studentErr } = await supabase
      .from("student_profiles")
      .select("id")
      .eq("user_id", req.user!.user_id);
    if (studentErr) return res.status(400).json({ detail: studentErr.message });
    if (!students || students.length === 0) {
      return res.status(400).json({ detail: "Complete your profile first" });
    }
    const student = students[0];
    const { data: active } = await supabase
      .from("credit_applications")
      .select("id")
      .eq("student_id", student.id)
      .in("status", ["pending", "under_review", "scoring_done", "matched", "contract_proposed", "active"]);
    if (active && active.length > 0) {
      return res.status(400).json({ detail: "You already have an active application" });
    }
    const { data: result, error } = await supabase
      .from("credit_applications")
      .insert({
        student_id: student.id,
        requested_monthly_amount: body.requested_monthly_amount,
        requested_duration_months: body.requested_duration_months,
        justification: body.justification,
        status: "pending",
      })
      .select();
    if (error) return res.status(400).json({ detail: error.message });
    return res.json({ success: true, data: result ? result[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.get("/credit-request", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data: students } = await supabase
      .from("student_profiles")
      .select("id")
      .eq("user_id", req.user!.user_id);
    if (!students || students.length === 0) {
      return res.json({ success: true, data: null });
    }
    const { data } = await supabase
      .from("credit_applications")
      .select("*")
      .eq("student_id", students[0].id)
      .order("created_at", { ascending: false })
      .limit(1);
    return res.json({ success: true, data: data && data.length > 0 ? data[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.get("/documents", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data, error } = await supabase
      .from("documents")
      .select("*")
      .eq("uploaded_by", req.user!.user_id);
    if (error) return res.status(400).json({ detail: error.message });
    return res.json({ success: true, data: data || [] });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.get("/contract", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data: students } = await supabase
      .from("student_profiles")
      .select("id")
      .eq("user_id", req.user!.user_id);
    if (!students || students.length === 0) {
      return res.json({ success: true, data: null });
    }
    const { data } = await supabase
      .from("contracts")
      .select("*")
      .eq("student_id", students[0].id)
      .order("created_at", { ascending: false })
      .limit(1);
    return res.json({ success: true, data: data && data.length > 0 ? data[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.post("/contract/:id/accept", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data, error } = await supabase
      .from("contracts")
      .update({ status: "accepted", accepted_at: new Date().toISOString() })
      .eq("id", req.params.id)
      .select();
    if (error) return res.status(400).json({ detail: error.message });
    return res.json({ success: true, data: data ? data[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.post("/contract/:id/refuse", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data, error } = await supabase
      .from("contracts")
      .update({ status: "terminated" })
      .eq("id", req.params.id)
      .select();
    if (error) return res.status(400).json({ detail: error.message });
    if (data && data.length > 0) {
      await supabase
        .from("credit_applications")
        .update({ status: "contract_refused" })
        .eq("id", data[0].application_id);
    }
    return res.json({ success: true, data: data ? data[0] : null });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.get("/payments", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data: students } = await supabase
      .from("student_profiles")
      .select("id")
      .eq("user_id", req.user!.user_id);
    if (!students || students.length === 0) {
      return res.json({ success: true, data: [] });
    }
    const { data: contracts } = await supabase
      .from("contracts")
      .select("id")
      .eq("student_id", students[0].id);
    const ids = contracts ? contracts.map((c) => c.id) : [];
    if (ids.length === 0) return res.json({ success: true, data: [] });
    const { data } = await supabase
      .from("payments")
      .select("*")
      .in("contract_id", ids)
      .order("due_date", { ascending: false });
    return res.json({ success: true, data: data || [] });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

router.get("/notifications", verifyToken, async (req: Request, res: Response) => {
  try {
    const { data, error } = await supabase
      .from("notifications")
      .select("*")
      .eq("user_id", req.user!.user_id)
      .order("created_at", { ascending: false })
      .limit(20);
    if (error) return res.status(400).json({ detail: error.message });
    return res.json({ success: true, data: data || [] });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

export default router;
